src/handlers/RF24StreamLogHandler.cpp

```diff
@@ -0,0 +1,217 @@
+{
+   if(logLevel > this->logLevel)
+   {
+      return;
+   }
+
+   appendTimestamp();
+   appendLogLevel(logLevel);
+   appendVendorId(vendorId);
+
+   // print formatted message
+   appendFormattedMessage(message, args);
+   stream->println("");
+}
+
+void RF24StreamLogHandler::setLogLevel(uint8_t logLevel)
+{
+   this->logLevel = logLevel;
+}
+
+void RF24StreamLogHandler::appendFormattedMessage(const char *format, va_list *args)
+{
+   for (; *format != 0; ++format)
+   {
+      if (*format == '%')
+      {
+         ++format;
+         appendFormat(*format, args);
+      }
+      else
+      {
+         stream->print(*format);
+      }
+   }
+}
+
+void RF24StreamLogHandler::appendFormattedMessage(const __FlashStringHelper *format, va_list *args)
+{
+   PGM_P p = reinterpret_cast<PGM_P>(format);
+   char c = pgm_read_byte(p++);
+   for (; c != 0; c = pgm_read_byte(p++))
+   {
+      if (c == '%')
+      {
+         c = pgm_read_byte(p++);
+         appendFormat(c, args);
+      }
+      else
+      {
+         stream->print(c);
+      }
+   }
+}
+
+void RF24StreamLogHandler::appendFormat(const char format, va_list *args)
+{
+   if (format == 's')
+   {
+      // print text from RAM
+      register char *s = (char *)va_arg(*args, int);
+      stream->print(s);
+
+      return;
+   }
+
+   if (format == 'S')
+   {
+      // print text from FLASH
+      register __FlashStringHelper *s = (__FlashStringHelper *)va_arg(*args, int);
+      stream->print(s);
+
+      return;
+   }
+
+   if (format == 'D' || format == 'F')
+   {
+      // print as double
+      stream->print(va_arg(*args, double));
+
+      return;
+   }
+
+   if (format == 'd' || format == 'i')
+   {
+      // print as integer
+      stream->print(va_arg(*args, int), DEC);
+
+      return;
+   }
+
+   stream->print(format);
+}
+
+void RF24StreamLogHandler::appendTimestamp()
+{
+   char c[12];
+   sprintf(c, "%10lu ", millis());
+   stream->print(c);
+   stream->print(" ");
+}
+
+void RF24StreamLogHandler::appendLogLevel(uint8_t logLevel)
+{
+   uint8_t logMainLevel = logLevel & 0b11111000;
+   uint8_t subLevel = logLevel & 0b00000111;
+
+   switch(logMainLevel)
+   {
+      case RF24LogLevel::ERROR:
+      {
+         stream->print(rf24LogLevels[0]);
+         break;
+      }
+      case RF24LogLevel::WARN:
+      {
+         stream->print(rf24LogLevels[1]);
+         break;
+      }
+      case RF24LogLevel::INFO:
+      {
+         stream->print(rf24LogLevels[2]);
+         break;
+      }
+      case RF24LogLevel::DEBUG:
+      {
+         stream->print(rf24LogLevels[3]);
+         break;
+      }
+      case RF24LogLevel::TRACE:
+      {
+         stream->print(rf24LogLevels[4]);
+         break;
+      }
+      default:
+      {
+         // unknown
+         stream->print("-----   ");
+         return;
+      }
+   }
+
+   if(subLevel == 0)
+   {
+      stream->print("   ");
+   } else
+   {
+      stream->print(":");
+      stream->print(subLevel);

```

Comment:
Print a sublevel after ```:``` but only when sublevel is different than 0.

Reply:
Does this condition still need to be checked? Now that all sublevels will be output as "Lvl xx". I think its ok if we skip outputting the parent level (for logging a sublevel) since it would require the user to understand the use of sublevels (demonstrated in examples) before invoking them.

Reply:
Currently the log looks like below:
```
        0  Lvl   0 RF24LogExample This is a log message with level 0
         6  Lvl   1 RF24LogExample This is a log message with level 1
        12  Lvl   2 RF24LogExample This is a log message with level 2
        18  Lvl   3 RF24LogExample This is a log message with level 3
        24  Lvl   4 RF24LogExample This is a log message with level 4
        30  Lvl   5 RF24LogExample This is a log message with level 5
        36  Lvl   6 RF24LogExample This is a log message with level 6
        41  Lvl   7 RF24LogExample This is a log message with level 7
        48  ERROR   RF24LogExample This is a log message with level 8
        54  ERROR:1 RF24LogExample This is a log message with level 9
        60  ERROR:2 RF24LogExample This is a log message with level 10
        66  ERROR:3 RF24LogExample This is a log message with level 11
        72  ERROR:4 RF24LogExample This is a log message with level 12
        78  ERROR:5 RF24LogExample This is a log message with level 13
        84  ERROR:6 RF24LogExample This is a log message with level 14
        91  ERROR:7 RF24LogExample This is a log message with level 15
        97   WARN   RF24LogExample This is a log message with level 16
       103   WARN:1 RF24LogExample This is a log message with level 17
       109   WARN:2 RF24LogExample This is a log message with level 18
       115   WARN:3 RF24LogExample This is a log message with level 19
       121   WARN:4 RF24LogExample This is a log message with level 20
       128   WARN:5 RF24LogExample This is a log message with level 21
       134   WARN:6 RF24LogExample This is a log message with level 22
       140   WARN:7 RF24LogExample This is a log message with level 23
       146   INFO   RF24LogExample This is a log message with level 24
       152   INFO:1 RF24LogExample This is a log message with level 25
       158   INFO:2 RF24LogExample This is a log message with level 26
       164   INFO:3 RF24LogExample This is a log message with level 27
       171   INFO:4 RF24LogExample This is a log message with level 28
       177   INFO:5 RF24LogExample This is a log message with level 29
       183   INFO:6 RF24LogExample This is a log message with level 30
       189   INFO:7 RF24LogExample This is a log message with level 31
       195  DEBUG   RF24LogExample This is a log message with level 32
       201  DEBUG:1 RF24LogExample This is a log message with level 33
       207  DEBUG:2 RF24LogExample This is a log message with level 34
       214  DEBUG:3 RF24LogExample This is a log message with level 35
       220  DEBUG:4 RF24LogExample This is a log message with level 36
       226  DEBUG:5 RF24LogExample This is a log message with level 37
       232  DEBUG:6 RF24LogExample This is a log message with level 38
       237  DEBUG:7 RF24LogExample This is a log message with level 39
       243  TRACE   RF24LogExample This is a log message with level 40
       249  TRACE:1 RF24LogExample This is a log message with level 41
       256  TRACE:2 RF24LogExample This is a log message with level 42
       262  TRACE:3 RF24LogExample This is a log message with level 43
       268  TRACE:4 RF24LogExample This is a log message with level 44
       274  TRACE:5 RF24LogExample This is a log message with level 45
       280  TRACE:6 RF24LogExample This is a log message with level 46
       286  TRACE:7 RF24LogExample This is a log message with level 47
       292  Lvl  48 RF24LogExample This is a log message with level 48
       299  Lvl  49 RF24LogExample This is a log message with level 49
       305  Lvl  50 RF24LogExample This is a log message with level 50
       311  Lvl  51 RF24LogExample This is a log message with level 51
       317  Lvl  52 RF24LogExample This is a log message with level 52
       323  Lvl  53 RF24LogExample This is a log message with level 53
       329  Lvl  54 RF24LogExample This is a log message with level 54
       335  Lvl  55 RF24LogExample This is a log message with level 55
       342  Lvl  56 RF24LogExample This is a log message with level 56
       348  Lvl  57 RF24LogExample This is a log message with level 57
       354  Lvl  58 RF24LogExample This is a log message with level 58
   ...
```
so for example all sublevels of ```ERROR``` are displayed as ```ERROR:x```. Keep in mind: there is only 8 sublevels for each main level. Other levels are shown as ```Lvl xxx```. This is just my interpretation. It can be changed, no problem.

Reply:
oh ok. I guess that's fine. It's just not what I'm used to seeing from the python lib. This would make more sense because we're not using the decimal system here (not using `level % 10`) to determine the parent level. Actually I think you _almost_ could say we're using an octal system (at least for sublevels).